Add XmlLicenseTransform E2E test (urn:mpeg:mpeg21:2003:01-REL-R-NS:licenseTransform)

krwq opened the pull request:
Last item in https://github.com/dotnet/corefx/issues/16781

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseEncryptedRef.cs:
```diff
@@ -0,0 +1,229 @@
+{
+    public class XmlLicenseEncryptedRef : IRelDecryptor
+    {
+        List<AsymmetricAlgorithm> asymmetricKeys = new List<AsymmetricAlgorithm>();
```
Style: Name should be `_asymmetricKeys`

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseEncryptedRef.cs:
```diff
@@ -0,0 +1,229 @@
+            asymmetricKeys.Add(key);
+        }
+
+        private static bool ArrayEqual(byte[] a, byte[] b)
```
You already have `using System.Linq;`; why not just use the SequenceEqual extension method, given that this is an early aborter?

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseEncryptedRef.cs:
```diff
@@ -0,0 +1,229 @@
+                            break;
+                        }
+                        else
+                            Assert.True(false, "Invalid License - MalformedKeyInfoClause");
```
Style: Block statements require braces when their sibling have braces, or they're close to existing braces.  (Two reasons this needs it)

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseEncryptedRef.cs:
```diff
@@ -0,0 +1,229 @@
+                            throw new CryptographicException("KeyDecryptionFailure");
+                        }
+
+                        aes = Aes.Create();
```
`using` this around the call to DecryptStream?

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseEncryptedRef.cs:
```diff
@@ -0,0 +1,229 @@
+                    Assert.True(false, "This test should not have KeyInfoName clauses");
+                }
+                else
+                    throw new CryptographicException("MalformedKeyInfoClause");
```
Style: Add braces since the sibling block statements had braces.

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseEncryptedRef.cs:
```diff
@@ -0,0 +1,229 @@
+            return DecryptStream(toDecrypt, aes);
+        }
+
+        private static Stream DecryptStream(Stream toDecrypt, SymmetricAlgorithm aes)
```
Style: It's weird to see `SymmetricAlgorithm aes`.  `alg`, perhaps?

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseEncryptedRef.cs:
```diff
@@ -0,0 +1,229 @@
+            toDecrypt.Read(encryptedContentValue, 0, encryptedContentValue.Length);
+
+            byte[] decryptedContent;
+            ICryptoTransform decryptor = aes.CreateDecryptor(aes.Key, IV);
```
`using` this.

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseEncryptedRef.cs:
```diff
@@ -0,0 +1,229 @@
+            using (var msDecrypt = new MemoryStream())
+            using (var csDecrypt = new CryptoStream(msDecrypt, decryptor, CryptoStreamMode.Write))
+            {
+                csDecrypt.Write(encryptedContentValue, 0, encryptedContentValue.Length);
```
Seems like you could avoid the encryptedContentValue array altogether by using `toDecrypt.CopyTo(csDecrypt)`

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseEncryptedRef.cs:
```diff
@@ -0,0 +1,229 @@
+                decryptedContent = msDecrypt.ToArray();
+            }
+
+            return new MemoryStream(decryptedContent);
```
Why? You... already have it in a memory stream.  You just need to set Position back to 0 (and not dispose it)

krwq replied:
@bartonjs CryptoStream is automatically disposing the underlying stream

bartonjs replied:
Can you use the overload which lets you keep the stream open?

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseEncryptedRef.cs:
```diff
@@ -0,0 +1,229 @@
+            using (var csDecrypt = new CryptoStream(msDecrypt, decryptor, CryptoStreamMode.Write))
+            {
+                csDecrypt.Write(encryptedContentValue, 0, encryptedContentValue.Length);
+                csDecrypt.FlushFinalBlock();
```
After you stop disposing msDecrypt and just return it, you can remove this call (because it's automatically called during Dispose)

krwq replied:
check the doc for CryptoStream - because the underlying stream is being disposed this can't be done differently

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseEncryptedRef.cs:
```diff
@@ -0,0 +1,229 @@
+            return new MemoryStream(decryptedContent);
+        }
+
+        public static void Encrypt(Stream toEncrypt, RSA key, out KeyInfo keyInfo, out EncryptionMethod encryptionMethod, out CipherData cipherData)
```
It's unfortunate when github requires scrolling, can you chop this?

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseTransformTest.cs:
```diff
@@ -30,6 +31,8 @@ public XmlNodeList UnprotectedGetInnerXml()
     public class XmlLicenseTransformTest
     {
+        public const string LicenseTransformUrl = "urn:mpeg:mpeg21:2003:01-REL-R-NS:licenseTransform";
+        public const string LicenseTransformNsUrl = "urn:mpeg:mpeg21:2003:01-REL-R-NS";
```
Why not define the (full) Url as NS + ":licenseTransform"?

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseTransformTest.cs:
```diff
@@ -83,6 +85,176 @@ public void OutputTypes()
+            }
+        }
+
+        static string GenerateLicenseXmlWithEncryptedGrants(RSA key, out string plainTextLicense)
```
Missing access modifier

bartonjs commented on src/System.Security.Cryptography.Xml/tests/XmlLicenseTransformTest.cs:
```diff
@@ -83,6 +85,176 @@ public void OutputTypes()
+                XmlDocument toDecrypt = LoadXmlWithLicenseNs(encryptedLicenseWithGrants, out nsManager);
+
+                var decryptor = new XmlLicenseEncryptedRef();
+                var transform = new XmlLicenseTransform()
```
Is this transform not ever used in SignedXml or EncryptedXml?

It seems like perhaps not, since it requires the extra Decryptor property-set.  Which sort of makes me wonder if it's something we should be porting.  But, I guess we've already done it, and done the work...

krwq replied:
@bartonjs it's in the spec and I believe it is mostly used indirectly

krwq commented:
@bartonjs just fixed the comments

krwq commented:
fixed merge conflict, simplified MemoryStream mess when using CryptoStream and did a tiny fix to line endings since git sometimes checks out files with "\r\n" on windows and the tests may fail locally for some people
